### PR TITLE
Bugfix FXIOS-9482 Reselect correct tab when undo close all tabs

### PR DESCRIPTION
**Pull request by ih-codes**
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-9482)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/20984)

## :bulb: Description
Each call to `removeTab()` updates the `backupCloseTab`. This PR saves the selected tab state prior to closing all tabs, then reassigns it to the `backupCloseTab` after the calls to `removeTab()` have been made.

Fix Demo:

https://github.com/user-attachments/assets/ae0cd1b0-fd05-4452-9165-3b9c56958bf0

## :pencil: Checklist
You have to check all boxes before merging
- [x] Filled in the above information (tickets numbers and description of your work)
- [x] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [ ] Wrote unit tests and/or ensured the tests suite is passing
- [ ] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [ ] If needed, I updated documentation / comments for complex code and public methods
- [ ] If needed, added a backport comment (example `@Mergifyio backport release/v120`)

### CODE REVIEW

File: firefox-ios/Client/TabManagement/Legacy/LegacyTabManager.swift [hunk 1]
```diff
@@ -637,15 +637,24 @@ class LegacyTabManager: NSObject, FeatureFlaggable, TabManager, TabEventHandler
...
+                                                isSelected: selectedTab?.tabUUID == tab.tabUUID)
         }
         backupCloseTabs = tabs
+
         for tab in currentModeTabs {
             await self.removeTab(tab.tabUUID)
         }
+
+        // Save the tab state that existed prior to removals (preserves original selected tab)
+        backupCloseTab = currentSelectedTab
```
Comment (cyndichin) [inline firefox-ios/Client/TabManagement/Legacy/LegacyTabManager.swift:656]: Should we add tests to check for this?

Comment (ih-codes) [reply firefox-ios/Client/TabManagement/Legacy/LegacyTabManager.swift:656]: Wouldn't be a bad idea! I just pushed up a couple tests.

It took me a while to find the right suite (`TabManagerTests.swift`). There do not appear to be many tab manager implementation tests as it is. Or did I miss a file as I was looking? 🤔

Comment (dataports) [reply firefox-ios/Client/TabManagement/Legacy/LegacyTabManager.swift:656]: Correct, there are not a lot of Tab Manager Implementation tests 😭 thanks for adding some, and thanks @cyndichin for advocating for it!

**Review by dataports — APPROVED** [review]
Looks great 👏

**Comment by dataports** [comment]
Gonna make some backports for this so it can go out in the dot release :)

**Comment by dataports** [comment]
@mergify backport release/v129

**Comment by dataports** [comment]
@mergify backport release/v128

File: firefox-ios/Client/TabManagement/Legacy/LegacyTabManager.swift (hunk 1 above)
Comment (cyndichin) [reply firefox-ios/Client/TabManagement/Legacy/LegacyTabManager.swift:656]: great thank you so much for writing up the tests so quickly!

**Comment by ih-codes** [comment]
After discussion with @cyndichin regarding some retain issues causing the new unit tests to fail, we are removing them from this PR and making a separate task to add those back and track down the underlying issues. Very likely having to do with background tasks dispatched during save sessions tate/preserving tabs.
[JIRA](https://mozilla-hub.atlassian.net/browse/FXIOS-9532) [GitHub](https://github.com/mozilla-mobile/firefox-ios/issues/21093)

**Pull request merged**